### ISSUE TITLE
Generate py3 module status spreadsheet

### ISSUE THREAD

**Issue by gundalow** [ansible/ansible#18456]
With the following columns in
* List of all modules
* Module description
* version_added
* Automated tests: if test/integration/targets/MODULE_NAME exists
* Manual tests
* Priority

**Comment by ansibot** [comment]
@gundalow, we need more information to resolve this issue. For that reason, we put this Issue into the 'needs_info' state.

Here are the required items we could not find in your description:
- issue type
 
When you have filled in the missing data, we will notify the module maintainer for further action.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
bot_skip

**Comment by mscherer** [comment]
I am not sure that we need the version_added, and the module description. That's already in the module.

I would however add the status somewhere:
- ported and tested extensively (or tested by the test suite)
- ported and tested manually
- ported, but not tested ( like automatic ) (not sure about this one)
- not tested
- not ported, need patch
- not ported, deps missing (for example, i would say "yum")

I wonder also if we could tag some of them as easyfix. For example, I expect the "expect" module to be a easy fix, as it doesn't requires much to be tested and or fixed.

Maybe we also need a list of associated bug ?

**Comment by frenzymadness** [comment]
@mscherer I like the "easyfix" idea.

**Comment by gundalow** [comment]
@mscherer Thanks for your suggestions.

I was thinking "version_added" could be useful to help us prioritise the list.
"Description" could be useful as we've got ~800 modules in there, some with non obvious prefixes or names.

"Easyfix" is a great idea, that will help the community a lot.

**Comment by mscherer** [comment]
mhh, how would we prioritize the list based on the version ?

And, so, maybe we should name the module using the full path ?

**Comment by gundalow** [comment]
Pull path is a good plan that should give enough context.
I was wondering about fixing older modules first. Just trying to work out how to prioritise 800 modules.

I wonder of easyfix can be based on the linecount (excluding docs)

**Comment by mscherer** [comment]
Yeah, easyfix could be done on linecount, or complexity

I would have placed priority based on remote modules (ie, something like service) vs something executed locally, but I am not sure which one should be first, ie, is there more demand for module that will manage a python 3 system, or for module being run locally on a python 3 controller, dealing with a remote API ?

I would say the 2nd, but we did mostly the 1st.

Older modules correlate also with being used by more people, so that's maybe a idea, indeed.

**Comment by frenzymadness** [comment]
Hello.

Is there some public draft of his sheet? AFAIK there is nothing I can help with related to Python 3 in GitHub project so I could help with modules porting.

**Comment by gundalow** [comment]
@frenzymadness Thanks for the offer of help.
Hi, no spreadsheet exists yet
@abadger Are you aware of any modules that need some Python 3 work?

**Comment by mscherer** [comment]
We still have https://github.com/ansible/ansible/projects/1 for python3 porting

**Comment by frenzymadness** [comment]
@mscherer Yes, we have, but AFAIK all issues/PRs there have a solution or are in *work in progress* state. So I am looking for another opportunity where I can help with Python 3 compatibility and it looks that modules porting is a good way.

If you know about something which is Python 3 related, add it please to mentioned project and/or let me know.

Or, maybe, I can try to investigate and make some automatic generator of a list of modules which aren't Python 3 compatible. I'll try it.

**Comment by frenzymadness** [comment]
I've created simple Python script based on `python-modernize` tool to check how many lines in every `*.py` would be changed by this tool to fix or improve compatibility and readability. I've run fixers from `python-modernize` one by one on every single Python file separately and I've created a matrix from result data. The matrix contains files on rows and fixers on columns and a number in the cells mean how many lines would `python-modernize` change.

There are two lists in spreadsheet - one with original data and one with data with removed lines and columns full of zeroes.

Result: https://docs.google.com/spreadsheets/d/1LY_wRrVs06_9ElGvpgV2xO6_9odmIZsLwTDkNqfl7FU/edit?usp=sharing

What do you think about it? Could it be valuable as an overview on Python 3 status of Ansible modules?

**Comment by abadger** [comment]
@frenzymadness yes indeed.  We need something like this.  Note that modernize goes above and beyond mere compatibility in some cases.  For instance, I noticed that fix_import shows up on practically all the mosules.  In most cases we don't need to add "from __future__ import absolute_import" to make code work and we can't add it until we branch devel for 2.4 (as we're keeping python-2.4 compatibility until then.)  We'll probably need to take a look at making some modules 100% clean in order to tell if there's other fixer tests that we should be ignoring for now.

Some other columns that could be nice: "Has automated tests"  => has integration tests that we're running in shippable.  "Has anyone tested this manually" => Someone in the community is using this module with python3 and it worked for their usages.

Might also be good to organize this so that anything that passed all tests is in a separate table from those that still need work.  As time goes on, it will be nice to see one table shrink and the other one grow.

**Comment by bcoca** [comment]
also .. remove any with /core/ or /extras/ in path, as those  modules should be dupes in the list

**Comment by frenzymadness** [comment]
@bcoca thanks, mentioned lines removed

@abadger Yes, this is the reason why the fixers are split one by one in the spreadsheet. For example `lib2to3.fixes.fix_ws_comma` fixes only readability by adding space after a comma in args list.

I added a new column where I generated a number of usages of a module in all `*.yml` files in the `test` folder. The number says how many occurrences of ` <modulename>:` are in all `*.yml` files in the `test` folder. I am not sure if this is a right approach to counting tests of a specific module, but it might be a good start.

The spreadsheet is now open for everybody so feel free to remove any fixer or add columns.

**Comment by mattclay** [comment]
@frenzymadness How much work would it be to change your script to provide a list of files and the fixes they need? If we had that, we could look at integrating it into our CI process to catch issues in PRs. Is that something you'd be interested in doing?

**Comment by frenzymadness** [comment]
@mattclay Sorry for the delay but I was on PyconSK and now I am facing troubles with bird flu (as a firefighter).

I can do it easily. Just send me output which you want that script to produce.  Another important part will be set up a list of the fixers to check with in CI.

List of all available fixers is here: http://python-modernize.readthedocs.io/en/latest/fixers.html

**Comment by mattclay** [comment]
@frenzymadness JSON output with the following information per issue would be ideal:

- path (relative path to file with issue, ex: `lib/ansible/modules/system/ping.py`)
- line (relevant line from path, if available)
- column (relevant column from path, if available)
- code (a code to uniquely identify the issue, if available)
- message (a brief one-line message explaining the issue to be corrected)

Alternatively, output in a pep8-like format would work:

```
path:line:column: code message
```

Being able to provide a list of fixers would be helpful. I'm guessing the fixers could be used as the codes associated with each message, if the messages themselves don't provide codes.

**Comment by gundalow** [comment]
@frenzymadness Hi, just reviewing PRs that are assigned to me. What can we do to help get this move along?

**Comment by frenzymadness** [comment]
@gundalow I am really sorry. I completely forget about this issue. I am gonna take a look on this ASAP and I'll send a link to script to discuss.

**Comment by frenzymadness** [comment]
@gundalow I updated my script based on the original lib2to3 module to fits your needs in CI. Output JSON contains path, code and simple message.

Line and column nums are not provided but code contains unified diff with changes made by python-modernize where line nums are included.

Unfortunately, I cannot provide a better message with applied fixer name because I have to apply all fixers at once. I tried to apply fixers one by one on each file but this approach has a huge impact on execution time.
For example, when I run this script in folder lib/ansible/modules/cloud with 399 Python files:
* All fixers at the same time - 20,48 seconds :)
* All fixers one by one - 13 mins and 20 seconds :(

Script: https://gist.github.com/frenzymadness/3ffac6f252a9f621b8f84366ec09ae54

Let me know if I can help more.
